conans/test/functional/graph/test_graph_build_mode.py

```diff
@@ -0,0 +1,167 @@
+import pytest
+import operator
+from conans.test.assets.genconanfile import GenConanfile
+from conans.test.utils.tools import TestClient
+
+
+class fakeop:
+    def __init__(self, function):
+        self.function = function
+    def __ror__(self, other):
+        return fakeop(lambda x, self=self, other=other: self.function(other, x))
+    def __or__(self, other):
+        return self.function(other)
+    def __rlshift__(self, other):
+        return fakeop(lambda x, self=self, other=other: self.function(other, x))
+    def __rshift__(self, other):
+        return self.function(other)
+    def __call__(self, value1, value2):
+        return self.function(value1, value2)
+
+
+@pytest.fixture(scope="module")
+def build_all():
+    """ Build a simple graph to test --build option
+
+        foobar <- bar <- foo
+               <--------|
+
+        All packages are built from sources to keep a cache.
+    :return: TestClient instance
+    """
+    client = TestClient()
+    client.save({"conanfile.py": GenConanfile().with_setting("build_type")})
+    client.run("export . foo/1.0@user/testing")
+    client.save({"conanfile.py": GenConanfile().with_require("foo/1.0@user/testing")
+                .with_setting("build_type")})
+    client.run("export . bar/1.0@user/testing")
+    client.save({"conanfile.py": GenConanfile().with_require("foo/1.0@user/testing")
+                .with_require("bar/1.0@user/testing")
+                .with_setting("build_type")})
+    client.run("export . foobar/1.0@user/testing")
+    client.run("install foobar/1.0@user/testing --build")
+
+    return client
+
+
+def test_install_build_single(build_all):
+    """ When only --build=<ref> is passed, only <ref> must be built
+    """
+    build_all.run("install foobar/1.0@user/testing --build=foo")
+
+    assert "bar/1.0@user/testing:7839863d5a059fc6579f28026763e1021268c55e - Cache" in build_all.out
+    assert "foo/1.0@user/testing:4024617540c4f240a6a5e8911b0de9ef38a11a72 - Build" in build_all.out
+    assert "foobar/1.0@user/testing:89636fbae346e3983af2dd63f2c5246505e74be7 - Cache" in build_all.out
+    assert "foo/1.0@user/testing: Forced build from source" in build_all.out
+    assert "bar/1.0@user/testing: Forced build from source" not in build_all.out
+    assert "foobar/1.0@user/testing: Forced build from source" not in build_all.out
+    assert "No package matching" not in build_all.out
+
+
+def test_install_build_double(build_all):
+    """ When both --build=<ref1> and --build=<ref2> are passed, only both should be built
+    """
+    build_all.run("install foobar/1.0@user/testing --build=foo --build=bar")
```

Comment:
Same here, isn't this covered elsewhere?

Reply:
Doing a quick search for --build on test folder, we don't have any other place which validate 2 packages with conan create command. We --build missing, but related to other features, like cross-compiling:


```
$ find conans/test/ -name "*.py" | xargs grep "\-\-build" conans/test/ -r | grep create --color=auto
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        t.run("create . --build missing")
conans/test/functional/generators/cmake_find_package_multi_test.py:        t.run("create . --build missing -s build_type=Release")
conans/test/functional/command/build_test.py:        client.run("create --build foo/1.0@user/stable . user/stable")
conans/test/functional/command/build_test.py:        client.run("create --build foo/1.0@user/stable --build bar/1.0@user/testing . user/testing")
conans/test/functional/build_helpers/msbuild_test.py:        client.run("create . Hello/1.2.1@lasote/stable --build")
conans/test/functional/build_helpers/msbuild_test.py:        client.run("create . Hello/1.2.1@lasote/stable --build -s arch=x86 -s build_type=Debug")
conans/test/functional/ui/json_output_test.py:        self.client.run("create . private_user/channel --build missing")
conans/test/integration/generators/package_info/package_info_test.py:        client.run("create conanfile_consumer.py consumer/1.0@us/ch --build missing")
conans/test/integration/conanfile/no_copy_source_test.py:        client.run("create . lasote/testing --build")
conans/test/integration/conanfile/no_copy_source_test.py:        client.run("create . lasote/testing --build")
conans/test/integration/conanfile/imports_test.py:        client.run("create . Pkg/0.1@user/testing --build=missing")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . user/testing --build missing")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/conanfile_extend_test.py:        client.run("create . conan/testing -o test:test_option=3 --build")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create conanfile.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=test4/0.1@ --build=test4 --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=app/0.1@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/0.1@ --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference app/0.1@ --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=app/0.1@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create variant.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/1.0@ --build=missing --lockfile-out=new.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create boost/conanfile.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create app/conanfile.py --profile:build=default --profile:host=default --build"
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create app/conanfile.py --profile:build=Windows --profile:host=Linux --build"
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=flac/1.0@ --lockfile-out=conan.lock --build")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --base --lockfile-out=base.lock "
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --lockfile=base.lock "
conans/test/integration/graph_lock/graph_lock_ci_test.py:        client.run("lock create --reference=PkgD/0.1@user/channel --build -pr=myprofile "
conans/test/integration/graph_lock/graph_lock_ci_test.py:        client.run("lock create pkgc/conanfile.py --name=pkgc --version=0.1 --build "
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create consumer.txt -pr=profile --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create --reference=LibC/1.0 --build=LibC --lockfile=libc.lock "
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create --reference=LibC/1.0 --build --lockfile=libc3.lock "
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("lock create --reference=nano/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("create ffmepg ffmpeg/1.0@ --build --lockfile=conan.lock")
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("lock create --reference=flac/1.0@ --lockfile-out=conan.lock --build")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . Pkg/0.1@user/testing -pr=myprofile --build=missing")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . MyLib/0.1@user/testing -pr=myprofile --build")
conans/test/integration/build_requires/profile_build_requires_test.py:        # now remove packages, ensure --build=missing also creates them
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . lasote/stable --profile ./profile.txt --build missing")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . lasote/stable --profile=./profile.txt --build=missing")
conans/test/integration/build_requires/build_requires_test.py:        client.run("create . app/0.1@lasote/stable --build=missing -pr=myprofile")
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client2.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        t.run("create package3.py pkg3/1.0@ --build=pkg1", assert_error=True)
conans/test/integration/package_id/package_id_requires_modes_test.py:        t.run("create package3.py pkg3/1.0@ --build=pkg1 --build=pkg3")
conans/test/integration/package_id/full_revision_mode_test.py:    client.run("create pkgb pkgb/0.1@ -pr=profile --build=missing")
conans/test/integration/conan_v2/generators/test_cpp_info.py:        t.run('create app.py app/version@ --build=missing', assert_error=True)
conans/test/integration/graph/transitive_deps_test.py:        self.t.run("create ./D user/channel --build=missing")
conans/test/integration/graph/version_range_override_test.py:        t.run("lock create conanfile.py --build -pr=myprofile")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build",
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/py_requires/python_requires_test.py:        client.run("create . project/1.0@user/test --build=missing")
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        client.run("create . user/channel --build missing")
conans/test/functional/generators/cmake_find_package_test.py:        t.run("create . --build missing")
conans/test/functional/generators/cmake_find_package_multi_test.py:        t.run("create . --build missing -s build_type=Release")
conans/test/functional/command/build_test.py:        client.run("create --build foo/1.0@user/stable . user/stable")
conans/test/functional/command/build_test.py:        client.run("create --build foo/1.0@user/stable --build bar/1.0@user/testing . user/testing")
conans/test/functional/build_helpers/msbuild_test.py:        client.run("create . Hello/1.2.1@lasote/stable --build")
conans/test/functional/build_helpers/msbuild_test.py:        client.run("create . Hello/1.2.1@lasote/stable --build -s arch=x86 -s build_type=Debug")
conans/test/functional/ui/json_output_test.py:        self.client.run("create . private_user/channel --build missing")
conans/test/integration/generators/package_info/package_info_test.py:        client.run("create conanfile_consumer.py consumer/1.0@us/ch --build missing")
conans/test/integration/conanfile/no_copy_source_test.py:        client.run("create . lasote/testing --build")
conans/test/integration/conanfile/no_copy_source_test.py:        client.run("create . lasote/testing --build")
conans/test/integration/conanfile/imports_test.py:        client.run("create . Pkg/0.1@user/testing --build=missing")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . user/testing --build missing")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/folders_access_test.py:        self.client.run("create . conan/stable --build")
conans/test/integration/conanfile/conanfile_extend_test.py:        client.run("create . conan/testing -o test:test_option=3 --build")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create conanfile.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=test4/0.1@ --build=test4 --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=app/0.1@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/0.1@ --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference app/0.1@ --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:            client.run("lock create --reference=app/0.1@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create variant.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/build_order_test.py:        client.run("lock create --reference=app/1.0@ --build=missing --lockfile-out=new.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create boost/conanfile.py --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create app/conanfile.py --profile:build=default --profile:host=default --build"
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        t.run("lock create app/conanfile.py --profile:build=Windows --profile:host=Linux --build"
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=flac/1.0@ --lockfile-out=conan.lock --build")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --base --lockfile-out=base.lock "
conans/test/integration/graph_lock/graph_lock_build_requires_test.py:        client.run("lock create --reference=pkg2/1.0@ --build --lockfile=base.lock "
conans/test/integration/graph_lock/graph_lock_ci_test.py:        client.run("lock create --reference=PkgD/0.1@user/channel --build -pr=myprofile "
conans/test/integration/graph_lock/graph_lock_ci_test.py:        client.run("lock create pkgc/conanfile.py --name=pkgc --version=0.1 --build "
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create consumer.txt -pr=profile --build=missing --lockfile-out=conan.lock")
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create --reference=LibC/1.0 --build=LibC --lockfile=libc.lock "
conans/test/integration/graph_lock/dynamic_test.py:        client.run("lock create --reference=LibC/1.0 --build --lockfile=libc3.lock "
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("lock create --reference=nano/1.0@ --build --lockfile-out=conan.lock")
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("create ffmepg ffmpeg/1.0@ --build --lockfile=conan.lock")
conans/test/integration/graph_lock/graph_lock_test.py:        client.run("lock create --reference=flac/1.0@ --lockfile-out=conan.lock --build")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . Pkg/0.1@user/testing -pr=myprofile --build=missing")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . MyLib/0.1@user/testing -pr=myprofile --build")
conans/test/integration/build_requires/profile_build_requires_test.py:        # now remove packages, ensure --build=missing also creates them
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . lasote/stable --profile ./profile.txt --build missing")
conans/test/integration/build_requires/profile_build_requires_test.py:        client.run("create . lasote/stable --profile=./profile.txt --build=missing")
conans/test/integration/build_requires/build_requires_test.py:        client.run("create . app/0.1@lasote/stable --build=missing -pr=myprofile")
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        client2.run('create . consumer/1.0@user/testing --build')
conans/test/integration/package_id/package_id_requires_modes_test.py:        t.run("create package3.py pkg3/1.0@ --build=pkg1", assert_error=True)
conans/test/integration/package_id/package_id_requires_modes_test.py:        t.run("create package3.py pkg3/1.0@ --build=pkg1 --build=pkg3")
conans/test/integration/package_id/full_revision_mode_test.py:    client.run("create pkgb pkgb/0.1@ -pr=profile --build=missing")
conans/test/integration/conan_v2/generators/test_cpp_info.py:        t.run('create app.py app/version@ --build=missing', assert_error=True)
conans/test/integration/graph/transitive_deps_test.py:        self.t.run("create ./D user/channel --build=missing")
conans/test/integration/graph/version_range_override_test.py:        t.run("lock create conanfile.py --build -pr=myprofile")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build",
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create gtest.py --profile=profile_host --profile:build=profile_build --build")
conans/test/integration/cross_building/build_requires_from_profile_test.py:        t.run("create library.py --profile:host=profile_host --profile:build=profile_build --build")
conans/test/integration/py_requires/python_requires_test.py:        client.run("create . project/1.0@user/test --build=missing")

```